src/util/treeToFlatArray.js

```diff
@@ -0,0 +1,243 @@
+    const showDistantAncestor = !(
+      childPathLength < cursor.length &&
+      !isCursorAncestor &&
+      !addDistantAncestorAndStop
+    )
+
+    const { isHidden, isPinned, isChildrenPinned, view, filteredChildren } = children.reduce((acc, child) => {
+      return {
+        isHidden: acc.isHidden || child.value === '=hidden',
+        isPinned: acc.isPinned || child.value === '=pin',
+        isChildrenPinned: acc.isChildrenPinned || child.value === '=pinChildren',
+        view: child.value === '=view' ? getFirstSubthoughtValue(unroot(childPath.concat(child))) : acc.view,
+        filteredChildren: acc.filteredChildren.concat(
+          !showHiddenThoughts && isFunction(child.value) ? [] : [child]
+        )
+      }
+    }, {
+      isHidden: false,
+      isPinned: false,
+      isChildrenPinned: false,
+      filteredChildren: []
+    })
+
+    const metaChildrenCount = children.length - filteredChildren.length
+
+    const isMeta = isFunction(value)
+
+    const isTableView = view === 'Table'
+
+    // hide if this node is itself a meta function or has children meta =hidden
+    // if showHiddenThoughts is true then don't hide at all
+    const shouldHide = !showHiddenThoughts && isHidden
+
+    const depthInfo = {
+      hiddenNodes: acc.depthInfo.hiddenNodes + (isHidden ? 1 : 0),
+      metaNodes: acc.depthInfo.metaNodes + (isMeta ? 1 : 0),
+    }
+
+    // do not recurse if hidden
+    if (!showDistantAncestor || shouldHide) return { ...acc, depthInfo }
+
+    // stop deeper recursion at certain depth where any descendant of cursor has more than one visible subthought
+    // stop further deeper recursion if max depth is reached
+    const stop =
+      (
+        (
+          viewInfo.table.column !== 1 &&
+          (addDistantAncestorAndStop || (isCursorDescendant && (viewInfo.table.column === 2 || visibleSiblingsCount > 1)))
+        ) &&
+        !isPinned &&
+        !pinChildren
+      ) ||
+      childPath.length - cursor.length === MAX_DEPTH_FROM_CURSOR
+
+    const distanceFromCursor = cursor.length - childPath.length
+
+    // if true the node will have reduced opacity on render
+    const isDistantThought =
+      (!isLeaf
+        ? distanceFromCursor >= 0
+        : distanceFromCursor >= (isCursorAncestor ? 2 : 1)) && !isCursor
+
+    const tableInfo = viewInfo.table
+
+    const { depthInfo: childrenDepthInfo, flatArray: flatArrayDescendants } = stop
+      ? { depthInfo: calculateDepthInfo(childPath, children), flatArray: [] } // stop recursion if stop is true (leaf nodes)
+      : getFlatArray({
+        startingPath: childPath,
+        children: filteredChildren,
+        cursor,
+        isLeaf,
+        showHiddenThoughts,
+        isParentCursorAncestor: isCursorAncestor,
+        isCursorDescendant: isCursorDescendant || isCursor,
+        visibleSiblingsCount: filteredChildren.length, // children nodes won't have to itearate its siblings
+        pinChildren: isChildrenPinned,
```

Comment:
Can we avoid passing `pinChildren` recursively, and calculate it at time of use since `attributeEquals` is `O(1)`? [L122](https://github.com/cybersemics/em/pull/503/files#diff-27b0c3c7a986c66d97df34fcf8c150acR122)

Reply:
Aah yes! I will change it.